README.md
stRange.js
==========
[![NPM version][npm-badge]](https://www.npmjs.com/package/strange)
[![Build status][travis-badge]](https://travis-ci.org/moll/js-strange)

stRange.js is a **range object** for JavaScript. Use it to have a single value
type with two endpoints and their boundaries. Also implements an interval tree
for quick lookups. Stringifies itself in the style of `[begin,end)` and allows
you to parse a string back. Also useful with PostgreSQL.

[npm-badge]: https://img.shields.io/npm/v/strange.svg
[travis-badge]: https://travis-ci.org/moll/js-strange.png?branch=master


Installing
----------
stRange.js follows [semantic versioning](http://semver.org/), so feel free to
depend on its major version with something like `>= 1.0.0 < 2` (a.k.a `^1.0.0`).

### Installing on Node.js
```
npm install strange
```

### Installing for the browser
stRange.js doesn't yet have a build ready for the browser, but you might be able
to use [Browserify][browserify] to have it run there till then.

[browserify]: https://github.com/substack/node-browserify


Using
-----
Create a Range object by passing in a beginning and end:
```javascript
var Range = require("strange")
var range = new Range(1, 5)
```

Check if something is a range and use it:
```javascript
var Range = require("strange")
if (range instanceof Range) console.log(range.begin, range.end)
```

### Bounds
You can set a range's bounds by passing the bounds as a two-character string of
parentheses as the 3rd argument:
```javascript
new Range(1, 3, "[)")
```

Bounds signify whether the range includes or excludes that particular endpoint.
The range above therefore includes numbers `>= 1 < 3`.

Pair | Meaning
-----|--------
`()` | open
`[]` | closed
`[)` | left-closed, right-open
`(]` | left-open, right-closed


### Parsing
To parse a range stringified by `Range.prototype.toString`, pass it to
`Range.parse`:

```javascript
Range.parse("[a,z)") // => new Range("a", "z", "[)")
```

To have stRange.js also parse the endpoints, pass a function to `Range.parse`:
```javascript
Range.parse("[42,69]", Number) // => new Range(42, 69)
```

### Using with PostgreSQL
The string format used by stRange.js matches [PostgreSQL's range type
format](http://www.postgresql.org/docs/9.4/static/rangetypes.html). You can
therefore use stRange.js to parse and stringify ranges for your database.


API
---
For extended documentation on all functions, please see the
[stRange.js API Documentation][api].

[api]: https://github.com/moll/js-strange/blob/master/doc/API.md

### [Range](https://github.com/moll/js-strange/blob/master/doc/API.md#Range)
- [begin](https://github.com/moll/js-strange/blob/master/doc/API.md#range.begin)
- [bounds](https://github.com/moll/js-strange/blob/master/doc/API.md#range.bounds)
- [end](https://github.com/moll/js-strange/blob/master/doc/API.md#range.end)
- [.prototype.compareBegin](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.compareBegin)(begin)
- [.prototype.compareEnd](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.compareEnd)(end)
- [.prototype.contains](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.contains)(value)
- [.prototype.intersects](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.intersects)(other)
- [.prototype.isBounded](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.isBounded)()
- [.prototype.isEmpty](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.isEmpty)()
- [.prototype.isFinite](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.isFinite)()
- [.prototype.isInfinite](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.isInfinite)()
- [.prototype.isUnbounded](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.isUnbounded)()
- [.prototype.toJSON](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.toJSON)()
- [.prototype.toString](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.toString)()
- [.prototype.valueOf](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.valueOf)()
- [.compareBeginToBegin](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.compareBeginToBegin)(a, b)
- [.compareBeginToEnd](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.compareBeginToEnd)(a, b)
- [.compareEndToEnd](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.compareEndToEnd)(a, b)
- [.parse](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.parse)(range, [parseEndpoint])
- [.union](https://github.com/moll/js-strange/blob/master/doc/API.md#Range.union)(union, a, b)

### [RangeTree](https://github.com/moll/js-strange/blob/master/doc/API.md#RangeTree)
- [.prototype.search](https://github.com/moll/js-strange/blob/master/doc/API.md#RangeTree.prototype.search)(valueOrRange)
- [.from](https://github.com/moll/js-strange/blob/master/doc/API.md#RangeTree.from)(ranges)


License
-------
stRange.js is released under a *Lesser GNU Affero General Public License*, which in summary means:

- You **can** use this program for **no cost**.
- You **can** use this program for **both personal and commercial reasons**.
- You **do not have to share your own program's code** which uses this program.
- You **have to share modifications** (e.g bug-fixes) you've made to this program.

For more convoluted language, see the `LICENSE` file.


About
-----
**[Andri Möll](http://themoll.com)** typed this and the code.  
[Monday Calendar](https://mondayapp.com) supported the engineering work.

If you find stRange.js needs improving, please don't hesitate to type to me now at [andri@dot.ee](mailto:andri@dot.ee) or [create an issue online](https://github.com/moll/js-strange/issues).


CHANGELOG.md
## 1.7.2 (Feb 22, 2017)
- Fixes `RangeTree` given a range that ended farther than ranges beginning after it.  
  This occurring was dependent on how the binary tree laid itself out.  
  Thanks, [Ross Allen](https://github.com/ssorallen), for reporting this!

## 1.7.1 (Jan 15, 2017)
- Fixes searching `RangeTree` with an empty range in it. Empty ranges are now ignored.

## 1.7.0 (Aug 5, 2016)
- Throws `RangeError` if the given bounds are not valid (not of the following: `[]`, `()`, `[)`, `(]`).  
  Thanks, [Nikhil Benesch](https://github.com/benesch), for the hint!

## 1.6.0 (Aug 5, 2016)
- Adds [`Range.compareBeginToEnd`][]
- Adds support for calling `RangeTree.prototype.search` with range to find ranges that intersect.

[`Range.compareBeginToEnd`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.compareBeginToEnd

## 1.5.0 (Jul 12, 2016)
- Adds [`Range.prototype.valueOf`][] to get a more primitive representation of a range.  
  Useful with [Egal.js][egal] or other libraries that compare value objects by their `valueOf` output.

[`Range.prototype.valueOf`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.valueOf
[egal]: https://github.com/moll/js-egal

## 1.4.0 (Jul 8, 2016)
- Makes `Range.prototype` a valid empty `Range`.  
  Allows you to use it as an empty range:

  ```javascript
  var EMPTY_RANGE = Range.prototype
  EMPTY_RANGE.isEmpty() // => true
  EMPTY_RANGE.contains(new Range(0, 1)) // => false
  ```

## 1.3.0 (Jul 17, 2015)
- Adds [`Range.compareBeginToBegin`][].
- Adds [`Range.compareEndToEnd`][].
- Adds [`Range.union`][].
- Adds [`Range.prototype.compareBegin`][].
- Adds [`Range.prototype.compareEnd`][].

- Sets numeric unbounded endpoints to `-Infinity`/`Infinity` if you pass
  `Number` to [`Range.parse`][] as the parse function.

  ```javascript
  Range.parse("[15,]", Number) // => new Range(15, Infinity)
  Range.parse("(,3.14]", Number) // => new Range(-Infinity, 3.14, "(]")
  ```

- Fixes [`Range.prototype.intersects`][] to handle exclusive unbounded ranges
  properly.

- Adds [`RangeTree`] for creating an interval tree (augmented binary search
  tree) for searching ranges that intersect with a given value.

[`Range.compareBeginToBegin`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.compareBeginToBegin
[`Range.compareEndToEnd`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.compareEndToEnd
[`Range.union`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.union
[`Range.prototype.compareBegin`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.compareBegin
[`Range.prototype.compareEnd`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.compareEnd
[`RangeTree`]: https://github.com/moll/js-strange/blob/master/doc/API.md#RangeTree

## 1.2.0 (Jul 4, 2015)
- Adds [`Range.prototype.isBounded`][].
- Adds [`Range.prototype.isUnbounded`][].
- Adds [`Range.prototype.isFinite`][].
- Adds [`Range.prototype.isInfinite`][].

[`Range.prototype.isBounded`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.isBounded
[`Range.prototype.isUnbounded`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.isUnbounded
[`Range.prototype.isFinite`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.isFinite
[`Range.prototype.isInfinite`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.isInfinite

## 1.1.0 (Jun 29, 2015)
- Adds [`Range.prototype.isEmpty`][].
- Adds [`Range.prototype.intersects`][].
- Adds [`Range.prototype.contains`][].

[`Range.prototype.isEmpty`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.isEmpty
[`Range.prototype.intersects`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.intersects
[`Range.prototype.contains`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.contains

## 1.0.0 (May 11, 2015)
- Adds support for setting range bounds with `new Range(1, 3, "[)")` to set
  a left-closed, right-open range.
- Adds [`Range.prototype.toString`][] to stringify a range.
- Adds [`Range.prototype.toJSON`][] as an alias to `Range.prototype.toString`.
- Adds [`Range.parse`][] to parse the stringified range.

[`Range.prototype.toString`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.toString
[`Range.prototype.toJSON`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.prototype.toJSON
[`Range.parse`]: https://github.com/moll/js-strange/blob/master/doc/API.md#Range.parse

## 0.1.337 (Oct 9, 2013)
- First release. Its future is an infinite exclusive range.


test/tree_test.js
var Range = require("..")
var RangeTree = require("../tree")

describe("RangeTree", function() {
  describe(".prototype.search", function() {
    describe("given value", function() {
      it("must return an empty array if empty", function() {
        var tree = RangeTree.from([])
        tree.search(42).must.be.empty()
      })

      it("must return an empty array if not found within 1", function() {
        var tree = RangeTree.from([new Range(0, 10)])
        tree.search(42).must.be.empty()
      })

      it("must return an empty array if with exclusive left boundary",
        function() {
        var tree = RangeTree.from([new Range(0, 10, "(]")])
        tree.search(0).must.be.empty()
      })

      it("must return an empty array if with exclusive right boundary",
        function() {
        var tree = RangeTree.from([new Range(0, 10, "[)")])
        tree.search(10).must.be.empty()
      })

      it("must return an empty array if not found within 3", function() {
        var ranges = [new Range(0, 10), new Range(15, 25), new Range(30, 40)]
        var tree = RangeTree.from(ranges)
        tree.search(42).must.be.empty()
      })

      it("must return range if within 1", function() {
        var tree = RangeTree.from([new Range(0, 10)])
        tree.search(5).must.eql([new Range(0, 10)])
      })

      it("must return range if within 2", function() {
        var tree = RangeTree.from([new Range(0, 10), new Range(20, 30)])
        tree.search(5).must.eql([new Range(0, 10)])
        tree.search(25).must.eql([new Range(20, 30)])
      })

      it("must return range if within 3", function() {
        var ranges = [new Range(0, 10), new Range(20, 30), new Range(40, 50)]
        var tree = RangeTree.from(ranges)
        tree.search(5).must.eql([new Range(0, 10)])
        tree.search(25).must.eql([new Range(20, 30)])
        tree.search(45).must.eql([new Range(40, 50)])
      })

      it("must return range if multiple with equal begin", function() {
        var tree = RangeTree.from([
          new Range(0, 20),
          new Range(10, 25),
          new Range(15, 35),
          new Range(15, 40),
        ])

        tree.search(37).must.eql([new Range(15, 40)])
      })

      it("must return multiple ranges if within 3", function() {
        var ranges = [new Range(0, 20), new Range(15, 35), new Range(30, 40)]
        var tree = RangeTree.from(ranges)
        tree.search(35).must.eql([new Range(15, 35), new Range(30, 40)])
      })

      it("must return multiple ranges with equal begin", function() {
        // Use plenty of ranges to ensure a naive dividing tree construction
        // puts the duplicate ranges further apart.
        var tree = RangeTree.from([
          new Range(0, 20),
          new Range(10, 25),
          new Range(15, 35),
          new Range(15, 40),
          new Range(35, 40),
          new Range(35, 50)
        ])

        tree.search(30).must.eql([new Range(15, 35), new Range(15, 40)])
      })

      it("must return multiple ranges if on two sides", function() {
        var tree = RangeTree.from([
          new Range(0, 15),
          new Range(5, 5),
          new Range(10, 25)
        ])

        tree.search(13).must.eql([new Range(0, 15), new Range(10, 25)])
      })

      it("must return multiple ranges if all equal", function() {
        var tree = RangeTree.from([new Range(5, 15), new Range(5, 15)])
        tree.search(10).must.eql([new Range(5, 15), new Range(5, 15)])
      })

      it("must return multiple ranges if one equal but exclusive", function() {
        var tree = RangeTree.from([
          new Range(5, 15),
          new Range(5, 10, "[)"),
          new Range(10, 15, "(]"),
          new Range(10, 20)
        ])

        tree.search(10).must.eql([new Range(5, 15), new Range(10, 20)])
      })

      it("must return multiple ranges ordered by ascending end", function() {
        var tree = RangeTree.from([
          new Range(0, 10),
          new Range(0, 15),
          new Range(0, 5)
        ])

        tree.search(5).must.eql([
          new Range(0, 5),
          new Range(0, 10),
          new Range(0, 15)
        ])
      })

      // This was a bug noticed on Jan 15, 2017 where an empty range caused
      // none to be matched.
      it("must return range if one empty", function() {
        var tree = RangeTree.from([new Range, new Range(0, 10)])
        tree.search(5).must.eql([new Range(0, 10)])
      })
    })

    describe("given range", function() {
      it("must return an empty array if empty", function() {
        var tree = RangeTree.from([])
        tree.search(new Range(42, 69)).must.be.empty()
      })

      it("must return an empty array if not found within 1", function() {
        var tree = RangeTree.from([new Range(0, 10)])
        tree.search(new Range(42, 69)).must.be.empty()
      })

      it("must return an empty array if with exclusive left boundary",
        function() {
        var tree = RangeTree.from([new Range(0, 10, "(]")])
        tree.search(new Range(-10, 0)).must.be.empty()
      })

      it("must return an empty array if with exclusive right boundary",
        function() {
        var tree = RangeTree.from([new Range(0, 10, "[)")])
        tree.search(new Range(10, 20)).must.be.empty()
      })

      it("must return an empty array if not found within 3", function() {
        var ranges = [new Range(0, 10), new Range(15, 25), new Range(30, 40)]
        var tree = RangeTree.from(ranges)
        tree.search(new Range(42, 69)).must.be.empty()
      })

      it("must return range if within 1", function() {
        var tree = RangeTree.from([new Range(0, 10)])
        tree.search(new Range(2, 7)).must.eql([new Range(0, 10)])
      })

      it("must return range if within 2", function() {
        var tree = RangeTree.from([new Range(0, 10), new Range(20, 30)])
        tree.search(new Range(2, 7)).must.eql([new Range(0, 10)])
        tree.search(new Range(22, 27)).must.eql([new Range(20, 30)])
      })

      it("must return range if within 3", function() {
        var ranges = [new Range(0, 10), new Range(20, 30), new Range(40, 50)]
        var tree = RangeTree.from(ranges)
        tree.search(new Range(2, 7)).must.eql([new Range(0, 10)])
        tree.search(new Range(22, 27)).must.eql([new Range(20, 30)])
        tree.search(new Range(42, 47)).must.eql([new Range(40, 50)])
      })

      it("must return range if multiple with equal begin", function() {
        var tree = RangeTree.from([
          new Range(0, 20),
          new Range(10, 25),
          new Range(15, 35),
          new Range(15, 40),
        ])

        tree.search(new Range(37, 39)).must.eql([new Range(15, 40)])
      })

      it("must return multiple ranges if within 3", function() {
        var ranges = [new Range(0, 20), new Range(15, 35), new Range(30, 40)]
        var tree = RangeTree.from(ranges)

        tree.search(new Range(32, 37)).must.eql([
          new Range(15, 35),
          new Range(30, 40)
        ])
      })

      it("must return multiple ranges with equal begin", function() {
        // Use plenty of ranges to ensure a naive dividing tree construction
        // puts the duplicate ranges further apart.
        var tree = RangeTree.from([
          new Range(0, 20),
          new Range(10, 25),
          new Range(15, 35),
          new Range(15, 40),
          new Range(35, 40),
          new Range(35, 50)
        ])

        tree.search(new Range(30, 32)).must.eql([
          new Range(15, 35),
          new Range(15, 40)
        ])
      })

      it("must return multiple ranges if on two sides", function() {
        var tree = RangeTree.from([
          new Range(0, 15),
          new Range(5, 5),
          new Range(10, 25)
        ])

        tree.search(new Range(6, 13)).must.eql([
          new Range(0, 15),
          new Range(10, 25)
        ])
      })

      it("must return multiple ranges if all equal", function() {
        var tree = RangeTree.from([new Range(5, 15), new Range(5, 15)])

        tree.search(new Range(7, 13)).must.eql([
          new Range(5, 15),
          new Range(5, 15)
        ])
      })

      it("must return multiple ranges if one equal but exclusive", function() {
        var tree = RangeTree.from([
          new Range(5, 15),
          new Range(5, 10, "[)"),
          new Range(10, 15, "(]"),
          new Range(10, 20)
        ])

        tree.search(new Range(10, 10)).must.eql([
          new Range(5, 15),
          new Range(10, 20)
        ])
      })

      it("must return multiple ranges ordered by ascending end", function() {
        var tree = RangeTree.from([
          new Range(0, 10),
          new Range(0, 15),
          new Range(0, 5)
        ])

        tree.search(new Range(3, 7)).must.eql([
          new Range(0, 5),
          new Range(0, 10),
          new Range(0, 15)
        ])
      })

      // This was a bug #2 (https://github.com/moll/js-strange/issues/2) where
      // a node's total range end was not set properly because it considered
      // the range(s) to the right of it to always end farther.
      it("must return range farther than later ranges", function() {
        var tree = RangeTree.from([
          new Range(-5, 5),
          new Range(0, 15),
          new Range(5, 10)
        ])

        tree.search(new Range(12, 13)).must.eql([new Range(0, 15)])
      })

      // The original bug report for the bug above had long timestamps that
      // I normalized below. Keeping them here as more insurance.
      it("must return range farther than later ranges in original bug report",
        function() {
        var tree = RangeTree.from([
          new Range(10, 20, "[)"),
          new Range(20, 30, "[)"),
          new Range(30, 40, "[)"),
          new Range(40, 50, "[)"),
          new Range(50, 60, "[)"),
          new Range(80, 100, "[)"),
          new Range(85, 86, "[)"),
        ])

        var matches = tree.search(new Range(90, 95, "[)"))
        matches.must.eql([new Range(80, 100, "[)")])
      })

      it("must not return ranges if exclusive", function() {
        var tree = RangeTree.from([
          new Range(0, 10, "[)"),
          new Range(20, 30, "(]")
        ])

        tree.search(new Range(10, 20)).must.eql([])
      })

      it("must not return ranges if exclusive and side by side", function() {
        var tree = RangeTree.from([new Range(0, 10, "[)")])
        tree.search(new Range(10, 15, "[)")).must.eql([])
      })
    })
  })
})


index.js
var INVALID_BOUNDS_ERR = "Invalid range bounds: "
module.exports = Range

/**
 * Create a new range object with the given begin and end endpoints.  
 * You can also pass a two character string for bounds. Defaults to` "[]"` for
 * an all inclusive range.
 *
 * You can use any value for endpoints. `Null` is considered infinity for
 * values that don't have a special infinity type like `Number` has `Infinity`.
 *
 * An empty range is one where either of the endpoints is `undefined` (like `new
 * Range`) or a range with two equivalent, but exculsive endpoints
 * (`new Range(5, 5, "[)")`).
 *
 * **Import**:
 * ```javascript
 * var Range = require("strange")
 * ```
 *
 * @example
 * new Range(10, 20) // => {begin: 10, end: 20, bounds: "[]"}
 * new Range(new Date(2000, 5, 18), new Date(2000, 5, 22))
 *
 * @class Range
 * @constructor
 * @param {Object} begin
 * @param {Object} end
 * @param {String} [bounds="[]"]
 */
function Range(begin, end, bounds) {
  if (!(this instanceof Range)) return new Range(begin, end, bounds)

  /**
   * Range's beginning, or left endpoint.
   *
   * @property {Object} begin
   */
  this.begin = begin

  /**
   * Range's end, or right endpoint.
   *
   * @property {Object} end
   */
  this.end = end

  /**
   * Range's bounds.
   *
   * Bounds signify whether the range includes or excludes that particular
   * endpoint.
   *
   * Pair | Meaning
   * -----|--------
   * `()` | open
   * `[]` | closed
   * `[)` | left-closed, right-open
   * `(]` | left-open, right-closed
   *
   * @example
   * new Range(1, 5).bounds // => "[]"
   * new Range(1, 5, "[)").bounds // => "[)"
   *
   * @property {String} bounds
   */
   this.bounds = bounds = bounds === undefined ? "[]" : bounds
   if (!isValidBounds(bounds)) throw new RangeError(INVALID_BOUNDS_ERR + bounds)
}

Range.prototype.begin = undefined
Range.prototype.end = undefined
Range.prototype.bounds = "[]"

/**
 * Compares this range's beginning with the given value.  
 * Returns `-1` if this range begins before the given value, `0` if they're
 * equal and `1` if this range begins after the given value.
 *
 * `null` is considered to signify negative infinity for non-numeric range
 * endpoints.
 *
 * @example
 * new Range(0, 10).compareBegin(5) // => -1
 * new Range(0, 10).compareBegin(0) // => 0
 * new Range(5, 10).compareBegin(0) // => 1
 * new Range(5, 10).compareBegin(null) // => 1
 *
 * @method compareBegin
 * @param {Object} begin
 */
Range.prototype.compareBegin = function(begin) {
  var a = this.begin === null ? -Infinity : this.begin
  var b = begin === null ? -Infinity : begin
  return compare(a, b) || (this.bounds[0] == "[" ? 0 : 1)
}

/**
 * Compares this range's end with the given value.  
 * Returns `-1` if this range ends before the given value, `0` if they're
 * equal and `1` if this range ends after the given value.
 *
 * `null` is considered to signify positive infinity for non-numeric range
 * endpoints.
 *
 * @example
 * new Range(0, 10).compareEnd(5) // => -1
 * new Range(0, 10).compareEnd(10) // => 0
 * new Range(0, 5).compareEnd(10) // => 1
 * new Range(0, 5).compareEnd(null) // => -1
 *
 * @method compareEnd
 * @param {Object} end
 */
Range.prototype.compareEnd = function(end) {
  var a = this.end === null ? Infinity : this.end
  var b = end === null ? Infinity : end
  return compare(a, b) || (this.bounds[1] == "]" ? 0 : -1)
}

/**
 * Check whether the range is empty.  
 * An empty range is one where either of the endpoints is `undefined` (like `new
 * Range`) or a range with two equivalent, but exculsive endpoints
 * (`new Range(5, 5, "[)")`).
 *
 * Equivalence is checked by using the `<` operators, so value objects will be
 * coerced into something comparable by JavaScript. That usually means calling
 * the object's `valueOf` function.
 *
 * @example
 * new Range().isEmpty() // => true
 * new Range(5, 5, "[)").isEmpty() // => true
 * new Range(1, 10).isEmpty() // => false
 *
 * @method isEmpty
 */
Range.prototype.isEmpty = function() {
  var a = this.begin === null ? -Infinity : this.begin
  var b = this.end === null ? Infinity : this.end
  if (a === undefined || b === undefined) return true
  return this.bounds != "[]" && compare(a, b) === 0
}

/**
 * Check whether the range is bounded.  
 * A bounded range is one where neither endpoint is `null` or `Infinity`. An
 * empty range is considered bounded.
 *
 * @example
 * new Range().isBounded() // => true
 * new Range(5, 5).isBounded() // => true
 * new Range(null, new Date(2000, 5, 18).isBounded() // => false
 * new Range(0, Infinity).isBounded() // => false
 * new Range(-Infinity, Infinity).isBounded() // => false
 *
 * @method isBounded
 */
Range.prototype.isBounded = function() {
  if (this.begin === undefined || this.end === undefined) return true
  return !(isInfinity(this.begin) || isInfinity(this.end))
}

/**
 * @method isFinite
 * @alias isBounded
 */
Range.prototype.isFinite = Range.prototype.isBounded

/**
 * Check whether the range is unbounded.  
 * An unbounded range is one where either endpoint is `null` or `Infinity`. An
 * empty range is not considered unbounded.
 *
 * @example
 * new Range().isUnbounded() // => false
 * new Range(5, 5).isUnbounded() // => false
 * new Range(null, new Date(2000, 5, 18).isUnbounded() // => true
 * new Range(0, Infinity).isUnbounded() // => true
 * new Range(-Infinity, Infinity).isUnbounded() // => true
 *
 * @method isUnbounded
 */
Range.prototype.isUnbounded = function() {
  return !this.isBounded()
}

/**
 * @method isInfinite
 * @alias isUnbounded
 */
Range.prototype.isInfinite = Range.prototype.isUnbounded

/**
 * Check if a given value is contained within this range.  
 * Returns `true` or `false`.
 *
 * @example
 * new Range(0, 10).contains(5) // => true
 * new Range(0, 10).contains(10) // => true
 * new Range(0, 10, "[)").contains(10) // => false
 *
 * @method contains
 * @param {Object} value
 */
Range.prototype.contains = function(value) {
  var a = this.begin
  var b = this.end

  return (
    (b === null || (this.bounds[1] === "]" ? value <= b : value < b)) &&
    (a === null || (this.bounds[0] === "[" ? a <= value : a < value))
  )
}

/**
 * Check if this range intersects with another.  
 * Returns `true` or `false`.
 *
 * Ranges that have common points intersect. Ranges that are consecutive and
 * with *inclusive* endpoints are also intersecting. An empty range will never
 * intersect.
 *
 * @example
 * new Range(0, 10).intersects(new Range(5, 7)) // => true
 * new Range(0, 10).intersects(new Range(10, 20)) // => true
 * new Range(0, 10, "[)").intersects(new Range(10, 20)) // => false
 * new Range(0, 10).intersects(new Range(20, 30)) // => false
 *
 * @method intersects
 * @param {Object} other
 */
Range.prototype.intersects = function(other) {
  if (this.isEmpty()) return false
  if (other.isEmpty()) return false

  return (
    Range.compareBeginToEnd(this, other) <= 0 &&
    Range.compareBeginToEnd(other, this) <= 0
  )
}

/**
 * Returns an array of the endpoints and bounds.
 *
 * Useful with [Egal.js](https://github.com/moll/js-egal) or other libraries
 * that compare value objects by their `valueOf` output.
 *
 * @example
 * new Range(1, 10, "[)").valueOf() // => [1, 10, "[)"]
 *
 * @method valueOf
 */
Range.prototype.valueOf = function() {
  return [this.begin, this.end, this.bounds]
}

/**
 * Stringifies a range in `[a,b]` format.
 *
 * This happens to match the string format used by [PostgreSQL's range type
 * format](http://www.postgresql.org/docs/9.4/static/rangetypes.html). You can
 * therefore use stRange.js to parse and stringify ranges for your database.
 *
 * @example
 * new Range(1, 5).toString() // => "[1,5]"
 * new Range(1, 10, "[)").toString() // => "[1,10)"
 *
 * @method toString
 */
Range.prototype.toString = function() {
  // FIXME: How to serialize an empty range with undefined endpoints?
  var a = stringify(this.begin)
  var b = stringify(this.end)
  return this.bounds[0] + a + "," + b + this.bounds[1]
}

/**
 * Stringifies the range when passing it to `JSON.stringify`.  
 * This way you don't need to manually call `toString` when stringifying.
 *
 * @example
 * JSON.stringify(new Range(1, 10)) // "\"[1,10]\""
 *
 * @method toJSON
 * @alias toString
 */
Range.prototype.toJSON = Range.prototype.toString
Range.prototype.inspect = Range.prototype.toString

/**
 * Compares two range's beginnings.  
 * Returns `-1` if `a` begins before `b` begins, `0` if they're equal and `1`
 * if `a` begins after `b`.
 *
 * @example
 * Range.compareBeginToBegin(new Range(0, 10), new Range(5, 15)) // => -1
 * Range.compareBeginToBegin(new Range(0, 10), new Range(0, 15)) // => 0
 * Range.compareBeginToBegin(new Range(0, 10), new Range(0, 15, "()")) // => 1
 *
 * @static
 * @method compareBeginToBegin
 * @param {Object} a
 * @param {Object} b
 */
Range.compareBeginToBegin = function(a, b) {
  var aBegin = a.begin === null ? -Infinity : a.begin
  var bBegin = b.begin === null ? -Infinity : b.begin
  if (a.bounds[0] === b.bounds[0]) return compare(aBegin, bBegin)
  else return compare(aBegin, bBegin) || (b.bounds[0] === "(" ? -1 : 1)
}

/**
 * Compares the first range's beginning to the second's end.  
 * Returns `<0` if `a` begins before `b` ends, `0` if one starts where the other
 * ends and `>1` if `a` begins after `b` ends.
 *
 * @example
 * Range.compareBeginToEnd(new Range(0, 10), new Range(0, 5)) // => -1
 * Range.compareBeginToEnd(new Range(0, 10), new Range(-10, 0)) // => 0
 * Range.compareBeginToEnd(new Range(0, 10), new Range(-10, -5)) // => 1
 *
 * @static
 * @method compareBeginToEnd
 * @param {Object} a
 * @param {Object} b
 */
Range.compareBeginToEnd = function(a, b) {
  var aBegin = a.begin === null ? -Infinity : a.begin
  var bEnd = b.end === null ? Infinity : b.end
  if (a.bounds[0] === "[" && b.bounds[1] === "]") return compare(aBegin, bEnd)
  else return compare(aBegin, bEnd) || 1
}

/**
 * Compares two range's endings.  
 * Returns `-1` if `a` ends before `b` ends, `0` if they're equal and `1`
 * if `a` ends after `b`.
 *
 * @example
 * Range.compareEndToEnd(new Range(0, 10), new Range(5, 15)) // => -1
 * Range.compareEndToEnd(new Range(0, 10), new Range(5, 10)) // => 0
 * Range.compareEndToEnd(new Range(0, 10), new Range(5, 10, "()")) // => 1
 *
 * @static
 * @method compareEndToEnd
 * @param {Object} a
 * @param {Object} b
 */
Range.compareEndToEnd = function(a, b) {
  var aEnd = a.end === null ? Infinity : a.end
  var bEnd = b.end === null ? Infinity : b.end
  if (a.bounds[1] === b.bounds[1]) return compare(aEnd, bEnd)
  else return compare(aEnd, bEnd) || (a.bounds[1] === ")" ? -1 : 1)
}

/**
 * Parses a string stringified by
 * [`Range.prototype.toString`](#Range.prototype.toString).
 *
 * To have it also parse the endpoints to something other than a string, pass
 * a function as the second argument.
 *
 * If you pass `Number` as the _parse_ function and the endpoints are
 * unbounded, they'll be set to `Infinity` for easier computation.
 *
 * @example
 * Range.parse("[a,z)") // => new Range("a", "z", "[)")
 * Range.parse("[42,69]", Number) // => new Range(42, 69)
 * Range.parse("[15,]", Number) // => new Range(15, Infinity)
 * Range.parse("(,3.14]", Number) // => new Range(-Infinity, 3.14, "(]")
 *
 * @static
 * @method parse
 * @param {String} range
 * @param {Function} [parseEndpoint]
 */
Range.parse = function(range, parse) {
  var endpoints = range.slice(1, -1).split(",", 2)
  var begin = endpoints[0] ? parse ? parse(endpoints[0]) : endpoints[0] : null
  var end = endpoints[1] ? parse ? parse(endpoints[1]) : endpoints[1] : null
  if (parse === Number && begin === null) begin = -Infinity
  if (parse === Number && end === null) end = Infinity
  return new Range(begin, end, range[0] + range[range.length - 1])
}

/**
 * Merges two ranges and returns a range that encompasses both of them.  
 * The ranges don't have to be intersecting.
 *
 * @example
 * Range.union(new Range(0, 5), new Range(5, 10)) // => new Range(0, 10)
 * Range.union(new Range(0, 10), new Range(5, 15)) // => new Range(0, 15)
 *
 * var a = new Range(-5, 0, "()")
 * var b = new Range(5, 10)
 * Range.union(a, b) // => new Range(-5, 10, "(]")
 *
 * @static
 * @method union
 * @param {String} union
 * @param {Range} a
 * @param {Range} b
 */
Range.union = function(a, b) {
  var aIsEmpty = a.isEmpty()
  var bIsEmpty = b.isEmpty()
  if (aIsEmpty && bIsEmpty) return new Range
  else if (aIsEmpty) return b
  else if (bIsEmpty) return a

  var begin = Range.compareBeginToBegin(a, b) <= 0 ? a : b
  var end = Range.compareEndToEnd(a, b) <= 0 ? b : a
  return new Range(begin.begin, end.end, begin.bounds[0] + end.bounds[1])
}

function isInfinity(value) {
  return value === null || value === Infinity || value === -Infinity
}

function isValidBounds(bounds) {
   switch (bounds) {
     case "()":
     case "[]":
     case "[)":
     case "(]": return true
     default: return false
   }
}

// The less-than operator ensures coercion with valueOf.
function compare(a, b) { return a < b ? -1 : b < a ? 1 : 0 }
function stringify(value) { return isInfinity(value) ? "" : String(value) }
